blackjax/inference/trajectory.py

```diff
@@ -379,55 +573,34 @@ def expand_once(expansion_state):
                 termination_state,
                 rate ** step,
                 step_size,
+                initial_energy,
             )
 
             left_trajectory, right_trajectory = reorder_trajectories(
                 direction, trajectory, new_trajectory
             )
 
-            # robust u-turn check when merging the two trajectory
-            # note this is different from the robust u-turn check done during
-            # trajectory building.
-            # momentum_sum_left = jax.tree_util.tree_multimap(
-            #     jnp.add,
-            #     left_trajectory.momentum_sum,
-            #     right_trajectory.leftmost_state.momentum,
-            # )
-            # is_turning_left = uturn_check_fn(
-            #     left_trajectory.leftmost_state.momentum,
-            #     right_trajectory.leftmost_state.momentum,
-            #     momentum_sum_left,
-            # )
-            # momentum_sum_right = jax.tree_util.tree_multimap(
-            #     jnp.add,
-            #     left_trajectory.rightmost_state.momentum,
-            #     right_trajectory.momentum_sum,
-            # )
-            # is_turning_right = uturn_check_fn(
-            #     left_trajectory.rightmost_state.momentum,
-            #     right_trajectory.rightmost_state.momentum,
-            #     momentum_sum_right,
-            # )
-
-            # merge the freshly integrated trajectory to the current trajectory
-            momentum_sum = jax.tree_util.tree_multimap(
-                jnp.add, left_trajectory.momentum_sum, right_trajectory.momentum_sum
-            )
-            merged_trajectory = Trajectory(
-                left_trajectory.leftmost_state,
-                right_trajectory.rightmost_state,
-                momentum_sum,
-                left_trajectory.num_states + right_trajectory.num_states,
-            )
+            merged_trajectory = merge_trajectories(left_trajectory, right_trajectory)
 
             # update the proposal
-            # we reject proposals coming from diverging or turning subtrajectories
+            # we reject proposals coming from diverging or turning subtrajectories,
+            # but accumulate average acceptance probabilty across entire trajectory
+            def update_cumsum_log_mh_accpet(inputs):
+                _, proposal, new_proposal = inputs
+                return Proposal(
+                    proposal.state,
+                    proposal.energy,
+                    proposal.weight,
+                    jnp.logaddexp(
+                        proposal.cumsum_log_mh_accpet, new_proposal.cumsum_log_mh_accpet
+                    ),
+                )
+
             sampled_proposal = jax.lax.cond(
-                is_diverging | has_terminated,
-                proposal,
-                lambda x: x,
-                (rng_key, proposal, new_proposal),
+                is_diverging | is_turning_subtree,
```

Comment:
strictly speaking `is_tuning_subtree` -> `has_terminated` to stay consistent with `termination_criterion`.